File: sem2/po/po_proj2/src/studia/prog_ob/projekt2/world_simulator/organisms/plants/plants_subclasses/PineBorscht.java
package studia.prog_ob.projekt2.world_simulator.organisms.plants.plants_subclasses;

import studia.prog_ob.projekt2.world_simulator.SimLogger;
import studia.prog_ob.projekt2.world_simulator.World;
import studia.prog_ob.projekt2.world_simulator.maps.MapLocation;
import studia.prog_ob.projekt2.world_simulator.organisms.Organism;
import studia.prog_ob.projekt2.world_simulator.organisms.plants.Plant;

public class PineBorscht extends Plant {
    public static final int PINEBORSCHT_SEEDING_THRESHOLD = BASE_PLANT_SEEDING_THRESHOLD;
    public static final int PINEBORSCHT_STRENGTH = 0;
    public static final int PINEBORSCHT_INITIATIVE = BASE_PLANT_INITIATIVE;
    public static final char PINEBORSCHT_SYMBOL = 'b';


    public PineBorscht(MapLocation new_map_location, World new_world) {
        super(PINEBORSCHT_STRENGTH, PINEBORSCHT_INITIATIVE, new_map_location, PINEBORSCHT_SYMBOL, new_world, PINEBORSCHT_SEEDING_THRESHOLD);
    }

    @Override
    public Organism spawn_offspring(MapLocation location) {
        return new PineBorscht(location, getWorld());
    }

    @Override
    public void collision(Organism invader) {
        SimLogger.log(invader + SimLogger.PINEBORSHT_POISONING + this);
        invader.kill_self();
        super.collision(invader);
    }

    @Override
    public void action() {
        kill_neighbours();
        super.action();
    }

    private int kill_neighbours() {
        Organism target = null;
        int kills = 0;
        for (int i = 1; i < 5; i++) {
            target = getWorld().getMap().get_neighbour(this, i);
            if (target != null && !target.is_plant()) {
                SimLogger.log(target + SimLogger.PINEBORSHT_POISONING + this);
                target.kill_self();
                kills++;
            }
        }
        return kills;
    }

    @Override
    public String toString() {
        return "Pine Borsht " + getMap_location() + " ";
    }
}

File: sem2/po/po_proj1/Plants/Plants_subclasses/Pine_Borscht.cpp
//
// Created by M Sz on 09/04/2023.
//

#include "Pine_Borscht.h"


Pine_Borscht::Pine_Borscht(int new_x, int new_y, World &new_world)
        : Plant(PINE_BORSCHT_STRENGTH, new_x, new_y, PINE_BORSCHT_SYMBOL, PINE_BORSCHT_SEEDING_THRESHOLD, new_world)
{}

Plant *Pine_Borscht::seed_new_plant(int spawnpoint_x, int spawnpoint_y) {
    return new Pine_Borscht(spawnpoint_x, spawnpoint_y, this->get_world());
}

void Pine_Borscht::action() {
    this->kill_neighbours();
    Plant::seed();
}


void Pine_Borscht::kill_neighbours() {
    Map& map = this->get_world().get_map();
    Organism* target = nullptr;
    XY this_location = this->get_location();

   target = this->get_world().get_map().get_neighbour(this_location, DIR_UP);
   if (target != nullptr && !target->is_plant())
       target->kill_self();

   target = this->get_world().get_map().get_neighbour(this_location, DIR_DOWN);
   if (target != nullptr && !target->is_plant())
       target->kill_self();

   target = this->get_world().get_map().get_neighbour(this_location, DIR_LEFT);
   if (target != nullptr && !target->is_plant())
       target->kill_self();

   target = this->get_world().get_map().get_neighbour(this_location, DIR_RIGHT);
   if (target != nullptr && !target->is_plant())
       target->kill_self();


}

std::string Pine_Borscht::login()
{
    return std::string("Pine Borscht");
}

void Pine_Borscht::collision(Organism *invading_organism) {
    invading_organism->kill_self();
}
